62_CAD/DrawResourcesFiller.cpp

```diff
@@ -780,154 +1957,528 @@ void DrawResourcesFiller::addLines_Internal(const CPolylineBase& polyline, const
+	const bool imageUsedForNextIntendedSubmit = (evicted.lastUsedFrameIndex == currentFrameIndex);
 
-		return true;
+	// NOTE: `deallocationWaitInfo` is crucial for both paths, we need to make sure we'll write to a descriptor arrayIndex when it's 100% done with previous usages.
+	if (imageUsedForNextIntendedSubmit)
+	{
+		// The evicted image is scheduled for use in the upcoming submit.
+		// To avoid rendering artifacts, we must flush the current draw queue now.
+		// After submission, we reset state so that data referencing the evicted slot can be re-uploaded.
+		submitDraws(intendedNextSubmit);
+		reset(); // resets everything, things referenced through mainObj and other shit will be pushed again through acquireXXX_SubmitIfNeeded
+
+		// Prepare wait info to defer index deallocation until the GPU has finished using the resource.
+		// we wait on the signal semaphore for the submit we just did above.
+		ISemaphore::SWaitInfo deallocationWaitInfo = { .semaphore = intendedNextSubmit.scratchSemaphore.semaphore, .value = intendedNextSubmit.scratchSemaphore.value };
+		suballocatedDescriptorSet->multi_deallocate(imagesArrayBinding, 1u, &evicted.arrayIndex, deallocationWaitInfo, &cleanupObject.get());
 	}
 	else
 	{
-		return false;
+		// The image is not used in the current frame, so we can deallocate without submitting any draws.
+		// Still wait on the semaphore to ensure past GPU usage is complete.
+		// TODO: We don't know which semaphore value the frame with `evicted.lastUsedFrameIndex` index was submitted with, so we wait for the worst case value conservatively, which is the immediate prev submit.
+		ISemaphore::SWaitInfo deallocationWaitInfo = { .semaphore = intendedNextSubmit.scratchSemaphore.semaphore, .value = intendedNextSubmit.scratchSemaphore.value };
+		suballocatedDescriptorSet->multi_deallocate(imagesArrayBinding, 1u, &evicted.arrayIndex, deallocationWaitInfo, &cleanupObject.get());
+	}
+}
+
+DrawResourcesFiller::ImageAllocateResults DrawResourcesFiller::tryCreateAndAllocateImage_SubmitIfNeeded(
+	const nbl::asset::IImage::SCreationParams& imageParams,
+	const asset::E_FORMAT imageViewFormatOverride,
+	nbl::video::SIntendedSubmitInfo& intendedNextSubmit,
+	std::string imageDebugName)
+{
+	ImageAllocateResults ret = {};
+
+	auto* device = m_utilities->getLogicalDevice();
+	auto* physDev = m_utilities->getLogicalDevice()->getPhysicalDevice();
+
+	bool alreadyBlockedForDeferredFrees = false;
+
+	// Attempt to create a GPU image and corresponding image view for this texture.
+	// If creation or memory allocation fails (likely due to VRAM exhaustion),
+	// we'll evict another texture from the LRU cache and retry until successful, or until only the currently-cachedImageRecord image remains.
+	while (imagesCache->size() > 0u)
+	{
+		// Try creating the image and allocating memory for it:
+		nbl::video::IGPUImage::SCreationParams params = {};
+		params = imageParams;
+		auto gpuImage = device->createImage(std::move(params));
```

Comment:
Where do you add the `ECF_MUTABLE_FORMAT_BIT` flag?  I think it should be added when override is valid and not equal to ICPUImage format (the image creation params given).
It shouldn't be user's job to add it

Reply:
Add the view format override in `params.viewFormat` (if valid and inequal to image format).
because we want the image creation to fail if we cannot create view with that format for the image.

Reply:
> Add the view format override in `params.viewFormat` (if valid and inequal to image format). because we want the image creation to fail if we cannot create view with that format for the image.

TODO:
https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/blob/dtm/62_CAD/DrawResourcesFiller.cpp#L2350-L2351